File: src/main/scala/com/leetcode/easy/LongestCommonPrefix.scala
package com.leetcode.easy

object LongestCommonPrefix {
  def longestCommonPrefix(strs: Array[String]): String = {
    if (strs.isEmpty) return ""
    
    val arrs = strs.sortBy(s => s.size).map(s => s.toCharArray)
    val len = arrs(0).size
    var res = Array[Char]()
    
    var i = 0
    while (i < len) {
      val c = arrs(0)(i)
      
      var j = 0
      while (j < arrs.size) {
        val c2 = arrs(j)(i)
        if (c != c2) return res.mkString
        j  = j + 1
      }
      
      res = res :+ c
      i = i + 1
    }
    res.mkString
  }
  
  def main(args: Array[String]): Unit = {
    println(longestCommonPrefix(Array()))
    println(longestCommonPrefix(Array("flower", "flow", "flight")))
    println(longestCommonPrefix(Array("dog", "racecar", "car")))
    println(longestCommonPrefix(Array("", "racecar", "car")))
  }
}

File: src/main/java/com/leecode/easy/LongestCommonPrefix.java
package com.leecode.easy;

public class LongestCommonPrefix {
    public String longestCommonPrefix(String[] strs) {
        if (strs == null || strs.length == 0) {
            return "";
        }
        int minLen = strs[0].length();
        for (int i = 1; i < strs.length; ++i) {
            minLen = Math.min(minLen, strs[i].length());
        }
        if (minLen == 0) {
            return "";
        }
        
        StringBuilder res = new StringBuilder();
        
        for (int i = 0; i < minLen; ++i) {
            char c = strs[0].charAt(i);
            boolean needBreak = false;
            for (int j = 1; j < strs.length; ++j) {
                if (strs[j].charAt(i) != c) {
                    needBreak = true;
                    break;
                }
            }
            if (needBreak) {
                break;
            }
            res.append(c);
        }
        return res.toString();
    }
    
    public static void main(String[] args) {
        LongestCommonPrefix app = new LongestCommonPrefix();
        System.out.println(app.longestCommonPrefix(new String[]{"aca", "cba"}));
        System.out.println(app.longestCommonPrefix(new String[]{"flower", "flow", "flight"}));
        System.out.println(app.longestCommonPrefix(new String[]{"dog", "racecar", "car"}));
    }
}